community/index/src/main/java/org/neo4j/index/gbptree/SeekCursor.java

```diff
@@ -178,299 +363,551 @@
-        this.pos = pos - 1;
-        this.keyCount = keyCount;
-    }
-
-    /**
-     * {@link TreeNode#keyCount(PageCursor) keyCount} is the only value read inside a do-shouldRetry loop
-     * which is used as data fed into another read. Because of that extra assertions are made around
-     * keyCount, both inside do-shouldRetry (requesting one more round in the loop) and outside
-     * (calling this method, which may throw exception).
-     *
-     * @param keyCount key count read from {@link TreeNode#keyCount(PageCursor)} and has "survived"
-     * a do-shouldRetry loop.
-     */
-    private boolean saneKeyCountRead( int keyCount )
-    {
-        return keyCount <= maxKeyCount;
-    }
-
-    @Override
-    public Hit<KEY,VALUE> get()
-    {
-        return this;
+        pos -= stride;
+        if ( !seekForward )
+        {
+            // The tree traversal is best effort when seeking backwards
+            // need to trigger search for key in next
+            concurrentWriteHappened = true;
+        }
     }
 
     @Override
     public boolean next() throws IOException
     {
         while ( true )
         {
-            pos++;
-            // some variables below are initialized to satisfy the compiler
-            byte nodeType;
-            long newGen = -1;
-            long rightSibling = -1;
-            long newGenGen = -1;
-            long rightSiblingGen = -1;
+            pos += stride;
+            // Read
             do
             {
-                nodeType = TreeNode.nodeType( cursor );
-                if ( nodeType != TreeNode.NODE_TYPE_TREE_NODE )
+                // Where we are
+                if ( !readHeader() )
                 {
-                    // If this node doesn't even look like a tree node then anything we read from it
-                    // will be just random data when looking at it as if it were a tree node.
                     continue;
                 }
 
-                currentNodeGen = bTreeNode.gen( cursor );
-                newGen = bTreeNode.newGen( cursor, stableGeneration, unstableGeneration );
-                keyCount = bTreeNode.keyCount( cursor );
-                if ( !keyCountIsSane() )
+                if ( verifyExpectedFirstAfterGoToNext )
                 {
-                    continue;
+                    pos = seekForward ? 0 : keyCount - 1;
+                    bTreeNode.keyAt( cursor,firstKeyInNode, pos );
                 }
 
-                if ( GenSafePointerPair.isSuccess( newGen ) )
+                if ( concurrentWriteHappened )
                 {
-                    newGenGen = bTreeNode.pointerGen( cursor, newGen );
+                    // Keys could have been moved so we need to make sure we are not missing any keys by
+                    // moving position back until we find previously returned key
+                    pos = searchKey( first ? fromInclusive : prevKey );
+                    if ( !seekForward && pos >= keyCount )
```

Comment:
Node: `[10 20 30 40]`, `keyCount = 4`
`pos = searchKey( 50 )` will result in `pos == 4`